core/include/gnuradio-4.0/TensorMath.hpp

```diff
@@ -0,0 +1,785 @@
+#ifndef GNURADIO_TENSOR_OPS_HPP
+#define GNURADIO_TENSOR_OPS_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <execution>
+#include <functional>
+#include <numeric>
+#include <type_traits>
+
+namespace gr::math {
+namespace config {
+
+// Size thresholds (can override with -DGR_MATH_TINY_THRESHOLD=value)
+#ifndef GR_MATH_TINY_THRESHOLD
+constexpr std::size_t TINY_THRESHOLD = 32;
+#else
+constexpr std::size_t TINY_THRESHOLD = GR_MATH_TINY_THRESHOLD;
+#endif
+
+#ifndef GR_MATH_MEDIUM_THRESHOLD
+constexpr std::size_t MEDIUM_THRESHOLD = 1024;
+#else
+constexpr std::size_t MEDIUM_THRESHOLD = GR_MATH_MEDIUM_THRESHOLD;
+#endif
+
+// SIMD Kernel Parameters - Architecture specific
+template<typename T>
+struct KernelParams;
+
+template<>
+struct KernelParams<float> {
+    static constexpr std::size_t MR         = 16;   // Rows in micro-kernel
+    static constexpr std::size_t NR         = 6;    // Cols in micro-kernel
+    static constexpr std::size_t SIMD_WIDTH = 8;    // AVX2: 8 floats
+    static constexpr std::size_t NC         = 4096; // L3 cache blocking
+    static constexpr std::size_t KC         = 256;  // Inner dim blocking
+    static constexpr std::size_t MC         = 128;  // L2 cache blocking
+};
+
+template<>
+struct KernelParams<double> {
+    static constexpr std::size_t MR         = 8;
+    static constexpr std::size_t NR         = 6;
+    static constexpr std::size_t SIMD_WIDTH = 4; // AVX2: 4 doubles
+    static constexpr std::size_t NC         = 4096;
+    static constexpr std::size_t KC         = 256;
+    static constexpr std::size_t MC         = 128;
+};
+
+template<typename T>
+requires std::integral<T>
+struct KernelParams<T> {
+    static constexpr std::size_t MR         = 4;
+    static constexpr std::size_t NR         = 4;
+    static constexpr std::size_t SIMD_WIDTH = 1; // integer types: smaller kernels (no SIMD for now)
+    static constexpr std::size_t NC         = 2048;
+    static constexpr std::size_t KC         = 256;
+    static constexpr std::size_t MC         = 64;
+};
+
+template<typename T>
+struct GemvParams {
+    static constexpr std::size_t SIMD_WIDTH  = KernelParams<T>::SIMD_WIDTH;
+    static constexpr std::size_t UNROLL      = 4;
+    static constexpr std::size_t CACHE_BLOCK = 2048;
+};
+
+} // namespace config
+
+enum class TransposeOp : std::int8_t {
+    NoTrans   = 0, /// no transpose
+    Trans     = 1, /// transpose
+    ConjTrans = 2, /// conjugated transpose
+    Dynamic   = -1 /// runtime decision (not yet implemented)
+};
+
+enum class MatrixSize : std::int8_t {
+    Tiny,   /// typ. 32x32
+    Medium, /// typ 4096x4096
+    Large   /// > Medium
+};
+
+struct cpu_policy {
+    static constexpr bool is_gpu       = false;
```

Comment:
Maybe an enum instead of bool - so that we can have other types of devices (maybe to follow opencl device types)?